### PR TITLE
Protein folding - chain lengths.

### PR DESCRIPTION
**Pull request by dlasecki**
### Summary

Solves #309.

### CODE REVIEW

File: qiskit_nature/problems/sampling/protein_folding/interactions/interaction.py
Comment (mrossinek) [inline qiskit_nature/problems/sampling/protein_folding/interactions/interaction.py:28]: This docstring (and probably others, too) will need updating

File: qiskit_nature/problems/sampling/protein_folding/interactions/random_interaction.py
Comment (mrossinek) [inline qiskit_nature/problems/sampling/protein_folding/interactions/random_interaction.py:24]: Do you want to address this as part of this PR or in the future?

File: qiskit_nature/problems/sampling/protein_folding/peptide/beads/base_bead.py
Comment (mrossinek) [inline qiskit_nature/problems/sampling/protein_folding/peptide/beads/base_bead.py:40]: Maybe write `an empty string` instead of `""`? Same elsewhere (presumably..)

File: qiskit_nature/problems/sampling/protein_folding/peptide/beads/base_bead.py
Comment (mrossinek) [inline qiskit_nature/problems/sampling/protein_folding/peptide/beads/base_bead.py:70]: I don't have a strong opinion on this, but `if self._residue_type` is probably more Pythonic than `!= ""`

**Review by mrossinek — COMMENTED** [review]
Thanks @dlasecki for the quick follow-up PR! :slightly_smiling_face: 
I did a quick review with some minor comments but in general this looks like a good approach!

File: qiskit_nature/problems/sampling/protein_folding/interactions/random_interaction.py
Comment (dlasecki) [reply qiskit_nature/problems/sampling/protein_folding/interactions/random_interaction.py:24]: Ideally we could solve it immediately, but I am not sure what is the best approach here.

Comment (pbark) [reply qiskit_nature/problems/sampling/protein_folding/interactions/random_interaction.py:24]: I have exactly the same question as @mrossinek  of the type change. In the previous version one would define the residues as `List[str]`. How can one define it in a single `Str` ?

File: qiskit_nature/problems/sampling/protein_folding/peptide/chains/main_chain.py
Comment (pbark) [inline qiskit_nature/problems/sampling/protein_folding/peptide/chains/main_chain.py:29]: Should this also be a `List(str)` or not?

File: qiskit_nature/problems/sampling/protein_folding/protein_folding_problem.py
Comment (pbark) [inline qiskit_nature/problems/sampling/protein_folding/protein_folding_problem.py:31]: would this also work with `["S","A","A","S",...]` ?

**Review by pbark — COMMENTED** [review]
I think this looks good. Just a small remark that I want to make sure that is properly addressed and then we can merge this.

File: qiskit_nature/problems/sampling/protein_folding/peptide/chains/main_chain.py
Comment (woodsp-ibm) [inline qiskit_nature/problems/sampling/protein_folding/peptide/chains/main_chain.py:106]:
```suggestion
        if main_chain_len != len(side_chain_residue_sequences):
            raise InvalidSizeException(
                f"The length of list of side chain residue sequences: "
```
I think the message should say `side` chain.... not `main` chain right

File: qiskit_nature/problems/sampling/protein_folding/peptide/chains/main_chain.py
Comment (woodsp-ibm) [inline qiskit_nature/problems/sampling/protein_folding/peptide/chains/main_chain.py:121]: Is this limitation of 1st, 2nd and last not being permitted a side chain exposed when you create a Peptide on its constructor. If I follow things down it seems that it just builds a MainChain and will fail if this is not the case. Looking at the example code, I guess I assumed that the absence of side chains in those places was just because it was an example, not that it was required as such. Is this something likely to change in the future? If not why would we want to user to pass these - would it make sense to define the side chain sequences as things that can be set - i.e length main chain - 3 so the first side chain in the sequence applies to main bead 3 (assuming indexes start at 1). Maybe passing them is clearer in that then each main has a side chain entry. But I think the behavior of 1st/2nd and last needs to be surfaced.

File: qiskit_nature/problems/sampling/protein_folding/peptide/chains/main_chain.py
Comment (mrossinek) [reply qiskit_nature/problems/sampling/protein_folding/peptide/chains/main_chain.py:29]: No that is the proposal of #309: rather than deal with a list of single-character strings, a simple string itself is much more reasonable. This should also resolve your comment below.

File: qiskit_nature/problems/sampling/protein_folding/peptide/chains/main_chain.py
Comment (mrossinek) [reply qiskit_nature/problems/sampling/protein_folding/peptide/chains/main_chain.py:121]: There are multiple aspects here:

1. 1st, 2nd and last not allowed: this is clearly documented in the Peptide's docstrings. The docstrings states that "this can be done without loss of generality" and refers to the paper for more information.

2. changing the side chain sequence: I don't think, that this is a good idea. Simply in terms of usability, I find it difficult to justify why a list `["a", "b", "c"]` would attach to some main sequence `"ABCDEF"` at the positions `a -> C`, `b -> D` and `c -> E`. That is both, not easy to document or explain and also just confusing. I would then rather see a good warning for the user.

3. Unittests: I am not sure if this exists already, but arguably there should be some unittests which assert the correct warnings being raised in edge cases similar to this.

File: qiskit_nature/problems/sampling/protein_folding/interactions/random_interaction.py
Comment (dlasecki) [reply qiskit_nature/problems/sampling/protein_folding/interactions/random_interaction.py:24]: Ok, for a moment I got confused and thought that a residue provided might potentially be coming from a main chain or a side chain. It might be the case in the future when extended to longer side chain but currently we only rely on the main chain. I will revert this change to avoid confusion for now.

File: qiskit_nature/problems/sampling/protein_folding/peptide/beads/base_bead.py
Comment (dlasecki) [reply qiskit_nature/problems/sampling/protein_folding/peptide/beads/base_bead.py:40]: I corrected in other places and somehow skipped this one. My bad.

File: qiskit_nature/problems/sampling/protein_folding/peptide/chains/main_chain.py
Comment (dlasecki) [reply qiskit_nature/problems/sampling/protein_folding/peptide/chains/main_chain.py:121]: I agree with Max and that was my motivation. 

Not crucial to this discussion but there is an issue to allow side chains on the second main bead at an expense of an additional qubit required for encoding the problem (#252) 

Thanks to Max I realized that my docs regarding this need to be fixed. I made too big of a shorthand in writing that actually result in a false claim.

File: qiskit_nature/problems/sampling/protein_folding/interactions/random_interaction.py
Comment (dlasecki) [reply qiskit_nature/problems/sampling/protein_folding/interactions/random_interaction.py:24]: @mrossinek The problem is that the random interaction does not use residues. It only needs the length of a residue. Ideally, it would only accept the length to avoid confusion that changing a residue would change the result somehow... 
But then it does not conform to the abstract method definition in the Interaction class...

But maybe it is not too bad...

Comment (mrossinek) [reply qiskit_nature/problems/sampling/protein_folding/interactions/random_interaction.py]: Yes I see. I think this is a very special case that the RandomInteraction does not make use of all the information given to it (i.e. it only cares about the length of the list and not its actual contents). This can happen and I don't think it's a problem.

And since we have a clear abstract definition, there is nothing we can do about it :upside_down_face: Simply keep the method definition, get the length from the list and ignore the rest. You can add a comment that this method does not care about the contents but other than that I would not change anything

File: qiskit_nature/problems/sampling/protein_folding/peptide/chains/main_chain.py
Comment (dlasecki) [reply qiskit_nature/problems/sampling/protein_folding/peptide/chains/main_chain.py]: We do have unit tests for that:
```
    def test_main_chain_illegal_side_chain_first(self):
        """Tests that an exception is thrown in case of illegal side chain."""
        main_chain_residue_seq = "SAAA"
        side_chain_residue_sequences = ["A", "", "S", ""]
        with self.assertRaises(InvalidSideChainException):
            _ = MainChain(
                main_chain_residue_seq,
                side_chain_residue_sequences,
            )

    def test_main_chain_illegal_side_chain_second(self):
        """Tests that an exception is thrown in case of illegal side chain."""
        main_chain_residue_seq = "SAAA"
        side_chain_residue_sequences = ["", "A", "S", ""]
        with self.assertRaises(InvalidSideChainException):
            _ = MainChain(
                main_chain_residue_seq,
                side_chain_residue_sequences,
            )

    def test_main_chain_illegal_side_chain_last(self):
        """Tests that an exception is thrown in case of illegal side chain."""
        main_chain_residue_seq = "SAAA"
        side_chain_residue_sequences = ["", "", "A", "S"]
        with self.assertRaises(InvalidSideChainException):
            _ = MainChain(
                main_chain_residue_seq,
                side_chain_residue_sequences,
            )
```

Comment (woodsp-ibm) [reply qiskit_nature/problems/sampling/protein_folding/peptide/chains/main_chain.py]: >  1st, 2nd and last not allowed: this is clearly documented in the Peptide's docstrings. The docstrings states that "this can be done without loss of generality" and refers to the paper for more information.

Where are you looking to see that - this is what I see in Peptide where I do not see the text you refer to
```
class Peptide:
    """A class containing information about beads and chains of a protein."""

    def __init__(
        self,
        main_chain_residue_sequence: str,
        side_chain_residue_sequences: List[str],
    ):
        """
        Args:
            main_chain_residue_sequence: String of characters that define residues for a main
                                        chain. Valid residue types are [A, C, D, E, F, G, H, I,
                                        K, L, M, N, P, Q, R, S, T, V, W, Y].
            side_chain_residue_sequences: List of characters that define residues for all side
                                        beads. Empty string if a side bead does not exist. Valid
                                        residue types are [A, C, D, E, F, G, H, I, K, L, M, N, P,
                                        Q, R, S, T, V, W, Y].
        """
```

Comment (mrossinek) [reply qiskit_nature/problems/sampling/protein_folding/peptide/chains/main_chain.py]: Sorry, it was indeed not in the Peptide but in the module-level docstring: https://github.com/Qiskit/qiskit-nature/blob/main/qiskit_nature/problems/sampling/protein_folding/__init__.py#L16-L18

Comment (woodsp-ibm) [reply qiskit_nature/problems/sampling/protein_folding/peptide/chains/main_chain.py]: Update: I found the text you refer to - it occurred to me after writing the above it might be in the overview text in the init file - and yes it was here qiskit_nature/problems/sampling/protein_folding/__init__.py

I still think PepTide class itself thought should make this clear when going to its API ref.

Comment (mrossinek) [reply qiskit_nature/problems/sampling/protein_folding/peptide/chains/main_chain.py]: Sounds good! @dlasecki could you copy that into the docstring of Peptide?

Comment (woodsp-ibm) [reply qiskit_nature/problems/sampling/protein_folding/peptide/chains/main_chain.py]: That comment sequence just above looks weird now - I tell you I figured it out after told me where it was! I had left the page open and perhaps it had not refreshed when I added the update after the prior comment - oh well!

Comment (dlasecki) [reply qiskit_nature/problems/sampling/protein_folding/peptide/chains/main_chain.py]: Do you mean duplicating what's in the init file to the Peptide docstring? Not sure about it... Maybe we could split it somehow and avoid duplication...

Comment (mrossinek) [reply qiskit_nature/problems/sampling/protein_folding/peptide/chains/main_chain.py]: I agree with Steve that this needs to go into the `Peptide` docstring. Whether we duplicate it on the module level or not is fine either way to me

Comment (woodsp-ibm) [reply qiskit_nature/problems/sampling/protein_folding/peptide/chains/main_chain.py]: If the more descriptive overall text stays in init, that would seem fine to me. I think here it should at least note the restriction on 1st, 2nd and last and perhaps link back to the main descriptive text for a higher level view/more information overall. It does not have to duplicate the entire text but there should be at least enough in PepTide to make them aware of the restriction/behavior if they first arrive at this class.

Comment (mrossinek) [reply qiskit_nature/problems/sampling/protein_folding/peptide/chains/main_chain.py]: Ah I did not want to say to copy the entire module level docstring. I would have only copied the sentence referring to the side chain restrictions.

Comment (dlasecki) [reply qiskit_nature/problems/sampling/protein_folding/peptide/chains/main_chain.py]: Updated the docs.

File: qiskit_nature/problems/sampling/protein_folding/__init__.py
Comment (mrossinek) [inline qiskit_nature/problems/sampling/protein_folding/__init__.py:16]:
```suggestion
main chain and optionally several side chains. Side chains cannot be attached to the first, second or
```
This will likely break `style` and `lint`...

File: qiskit_nature/problems/sampling/protein_folding/__init__.py
Comment (mrossinek) [inline qiskit_nature/problems/sampling/protein_folding/__init__.py]:
```suggestion
main bead allows for saving 1 qubit without loss of generality due to symmetry arguments (see
```

File: qiskit_nature/problems/sampling/protein_folding/__init__.py
Comment (mrossinek) [inline qiskit_nature/problems/sampling/protein_folding/__init__.py]: Just a general comment: right now, this is all one big paragraph. Maybe this will be more readable with some line breaks,

File: qiskit_nature/problems/sampling/protein_folding/residue_validator.py
Comment (mrossinek) [inline qiskit_nature/problems/sampling/protein_folding/residue_validator.py:20]: Why does this still need to support `List[str]`?

File: qiskit_nature/problems/sampling/protein_folding/peptide/chains/main_chain.py
Comment (mrossinek) [inline qiskit_nature/problems/sampling/protein_folding/peptide/chains/main_chain.py:73]: Since you changed it above, maybe do the same here. I am not sure about other occurrences.

**Comment by mrossinek** [comment]
Other than this comment I think this looks good to me :+1: 
You may want to merge `main` back in though, too

**Review by mrossinek — APPROVED** [review]
LGTM! Let's await another review before merging though :+1:

**Review by pbark — APPROVED** [review]
This also LGTM! Thanks for the quick fix!

**Pull request merged**
